### ISSUE TITLE
Tachiyomi volume downloads sometimes missing pages

### ISSUE THREAD

**Issue by noxiousnick** [Kareadita/Kavita#2494]
**If this is a feature request, request [here](https://feats.kavitareader.com/) instead. Feature requests will be deleted from Github.**

Please put as much information as possible to help me understand your issue. OS, browser, version are very important!

**Describe the bug**
When using Tachiyomi, I use the download button to save a local copy of a manga volume to my phone or tablet for offline reading later. Sometimes the downloaded volume is missing pages, and it seems to be random if it happens at all and which pages it happens to, but most often it is the cover or one of the first few pages. To make up for the missing pages, tachiyomi then shows some other pages duplicated, so that the total page count of the volume remains the same. For example, I would expect a volume to show pages in this order: Cover, page 1, page 2, page 3, page 4. But there's a random chance I will see this instead: Page 4, Page 5, Page 3, Page 3, Page 4. This happens on both my smartphone and my fire tablet, but happens much more often on my tablet.

**To Reproduce**
1. Connect Tachiyomi app to Kavita using the OPDS url
2. Pick any manga volume from the library and "download to local device"
3. Read the downloaded volume
4. Sometimes pages are missing/jumbled, it's random

**Expected behavior**
Volume should extract correctly every time without missing pages.

**Kavita Server:**
 - Raspberry Pi 4b
 - Docker image lscr.io/linuxserver/kavita:latest
 - Current version: 0.7.11.2

**Smartphone:**
 - Device: Galaxy S21
 - OS: Android
 - Tachiyomi 0.14.7
 - Kavita Extension 1.4.11

**Tablet:**
 - Device: Kindle Fire
 - OS: Fire OS
 - Tachiyomi 0.14.7
 - Kavita Extension 1.4.11

**Additional context**
I only just started using Tachiyomi about a half a year ago and this issue has always been present. I only just now decided to try and find a fix for this issue, and after reaching out in the discord I was suggested to create this issue as this apparently is not a well known issue.

I captured this error in the kavita logs while testing the issue recently:

```
[Kavita] [2023-12-11 10:27:14.658 -05:00  94] [Warning] API.Services.ArchiveService [ExtractArchive] There was a problem extracting /manga/Choujin X/Choujin X v04 (2023) (Digital) (1r0n).cbz to /app/kavita/config/cache/8242/
System.IO.IOException: The process cannot access the file '/app/kavita/config/cache/8242/Choujin X - c022 (v04) - p002 [VIZ Media] [Digital] [1r0n].png' because it is being used by another process.
```

Based on the error message, my layman assumption is Kavita is attempting to do multiple parallel things to the cbz file while tachiyomi is downloading it and Kavita is extracting it to its local cache at the same time. Sometimes there is a missing page from the cache due to the error, but more pages will be jumbled on Tachiyomi than just the one missing from cache.

While discussing this issue in the discord I was told this sounds related to a similar issue: #2478

**Comment by noxiousnick** [comment]
Some more testing I've done today that might help. Today I used the web browser on my computer to pre-load the volumes I wanted to download. My process went like this:
 - Navigate to manga series in the web browser on the computer
 - Select the volume I intend to download and read it incognito (just to preserve actual reading history)
 - Wait a few minutes for it to finish extracting the volume to cache
 - Open tachiyomi on my tablet and download the same volume

This gave me a 99% increase in reliability. I tested probably two dozen volumes, and only one volume gave me jumbled pages. It ended up reading like this: Page 2, Page 2, Page 4, Page 4, Page 5 ..... and so on correctly from there. I confirmed the volume cached correctly and just had to delete and redownload it once in tachiyomi to fix it.

This test is telling me that Tachiyomi is probably calling Kavita functions faster than it's ready to deliver them? Not sure how else to dig into this as a non-developer but I'm happy to do more testing if it helps to track this down and solve it.

**Comment by majora2007** [comment]
Thanks, I'll take a look into this next year.

**Comment by hstraten** [comment]
I have the same issue, it happens when doing a mass download (download everything unread) or when downloading a single chapter, no difference. Only workaround was indeed getting Kavita to place it in the cache by reading it in the webreader and then downloading it.
But I sometimes see the same media error there (seems to be related to issue #2604) so not limited to devices, happens in the webreader as well

I have sort of the same setup, but instead of a Pi I use an Intel NUC, and my tablet is a Boox Tab C mini
Since Tachiyomi development stopped I moved to Aniyomi 0.15.2.4 (kavita extention version 1.4.13) Kavita is @ 0.7.14
(recently moved to Mihon 1.16.5 Kavita plugin 1.4.15 and on the latest nightly at this moment 0.8.1.3)

If you can't reproduce the issue, I'm very willing to install any test version you throw at me, I can easily reproduce the issue.

**Comment by DeWir3** [comment]
I'm having the same issue and I too believe #2604 references the same thing. I'm mostly using Tachiyomi / Mihon to read my Manga and Comic from Kavita and it has been happening very frequently recently. Pages are skipped either right at the beginning or in the middle, and then other pages are repeated instead, usually the last page until the end of the chapter/issue. It seems to be happening semi-randomly but persistently across reloads (maybe some form of caching somehow?). It's also happening across platforms (Tachiyomi/Mihon on Android through the plugin, kavita webreader on Android, kavita webreader on Linux) and with different series in different formats (cbz and pdf).
The first time I noticed it, it was happening in Tachiyomi with chapters I downloaded into the app, but didn't happen when I "streamed" them without downloading. But it now happens in both cases, and also directly in the browser as I mentioned. Right now I can open the same manga chapter with the same account on my laptop's browser and it loads fine, open it on my tablet's browser through tailscale and it's missing/duplicating pages, open it in the same browser through the local IP and it works fine, and open it on my laptop again in the same browser through the tailscale IP and it is fine. 
But I have also seen this issue happen through the local IP on my laptop's browser a few days ago. In that instance it was happening with almost all chapters of a series I recently added. Then I removed the series from the library and added it back in, which seemed to fix it at least temporarily. But since it is happening across so many different scenarios, but also not consistently, I have no idea if it will stay that way.

My kavita server is currently on 0.8.1.0 running in the ~~official~~ linuxserver docker image on top of a NixOS server on an Intel NUC. Mihon is Version 0.16.5 with Kavita plugin 1.4.15 downloaded from [here](https://raw.githubusercontent.com/Kareadita/tachiyomi-extensions/repo/index.min.json) running on Android 13 and 14.

**Comment by majora2007** [comment]
This has been deployed to the nightlies. If you can, please try it out.

**Comment by hstraten** [comment]
Cleared all my downloaded comics from Mihon, well most of them, and started a complete new download for about 50 comics, never seen a media error again, and no pages in the wrong order either. Looks to me it's all good. Thanks you so much!

**Comment by majora2007** [comment]
Awesome to hear. Thanks for validating for me.

**Comment by DeWir3** [comment]
I realised I am using the linuxserver docker image and make use of its uid/gid setting. So switching to the nightly build would likely mess up the permissions in my library. hstraten's verdict sounds promising though and I look forward to giving it a proper shake once it hits stable. Thanks a bunch for all your work!

**Comment by DeWir3** [comment]
I did end up setting up a small testing environment to test this with the latest nightly. I downloaded one comic PDF with 88 pages and 6 manga chapters in cbz with 13-18 pages in a row into Mihon, none of them had the issue. Then I went back to my current stable setup, downloaded the same files into Mihon. The comic there had several missing / doubled pages, and 5 out of the 6 manga chapters had at least one missing / doubled pages. While it's just a small sample, the odds are definitely pointing to the issue being fixed! Thanks again to everyone for working on this, and @SamuelBMartins if you have some form of tip jar somewhere I'be happy to buy you a coffee for fixing this :)

**Comment by Hamuko** [comment]
I think I'm also having this issue on 0.8.1.0 (jvmilazz0/kavita / `afb2904c41006332f6bf9c5bca52990b6b76039ddace73c2fda4413c39972f28`). Some volumes are missing a bunch of pages when I open them either in Paperback or using the web UI.

I've noticed it happening randomly with different volumes/chapters, but the latest one that I have giving me issues shows me the cover (page 0), page 1 and page 35 as pages 1, 2 and 3 in Kavita web UI and when viewing through Paperback.

![Screenshot 2024-07-02 at 17 57 00](https://github.com/Kareadita/Kavita/assets/1270914/a9ebb9e8-d4d2-45e5-871d-4fc0d9d4e965)

I had it happen with the previous volume in this series too, where it skipped to like page 10, but somehow I managed to get the pages working correctly without doing anything to the files. I think what got it working was rescanning the series. And in fact, rescanning the series now has my example volume correct until page 9, after which things get really bizarre: page 9 in Kavita gives me page 41, pages 10 and 11 in Kavita give me pages 10 and 11, and page 12 in Kavita gives me page 125.

![Screenshot 2024-07-02 at 18 10 12](https://github.com/Kareadita/Kavita/assets/1270914/e52c3b44-7695-49ae-9410-0eed9d5358cb)
![Screenshot 2024-07-02 at 18 10 17](https://github.com/Kareadita/Kavita/assets/1270914/5995fc4e-b678-45e2-8767-5220eafc98b2)
![Screenshot 2024-07-02 at 18 10 29](https://github.com/Kareadita/Kavita/assets/1270914/bce85069-9859-4f2c-b875-8f8a8e7cb165)

Did a couple of more rescans after that and I think this volume is finally showing up correctly for me.

It's also not restricted to a single download source, since it's happened with manga downloaded from different places. All are CBZ (ZIP) archives though.

Should I switch to a nightly version of jvmilazz0/kavita?
